### PR TITLE
Initial editor demo implementation

### PR DESCRIPTION
**Pull request by rshipp**
Task: https://app.asana.com/0/1143269577537246/1147692631250862/f

Summary: Implement a barebones demo of inserting custom React components into an editor with balloon and block toolbars, and an HTML view. Edit/create functionality is restored.

Play around with the demo (as admin): https://platform-editor-feature-9acrvx.herokuapp.com/contents/new

(You can just collapse the ckeditor inspector on the bottom half of the page - that's there for debugging purposes only.)

This is basically pieces taken from CKEditor's [React widget demo](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html), and [balloon-block editor build](https://ckeditor.com/docs/ckeditor5/latest/builds/guides/overview.html#balloon-block-editor), and fit into our app.

### CODE REVIEW

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -11,6 +11,8 @@
  * It is generally better to create a new file per style scope.
  *
  */
+//*= require_tree .
```
Comment (rshipp) [inline app/assets/stylesheets/application.scss]: idk why this was taken out of the old application.scss but we need it here to include all the other scss files

File: app/assets/stylesheets/contents.scss [hunk 2]
```diff
@@ -1,3 +1,164 @@
 // Place all the styles related to the contents controller here.
-// They will automatically be included in application.css.
+// They will automatically be included in.application.css.
 // You can use Sass (SCSS) here: http://sass-lang.com/
+/* assets/styles.css */
```
Comment (rshipp) [inline app/assets/stylesheets/contents.scss]: css modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#styles-and-assets

File: app/javascript/ckeditor/contentpartpreviewediting.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import Plugin from '@ckeditor/ckeditor5-core/src/plugin';
```
Comment (rshipp) [inline app/javascript/ckeditor/contentpartpreviewediting.js]: js modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#productprevieweditingjs

File: app/javascript/ckeditor/insertcontentpartpreviewcommand.js [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import Command from '@ckeditor/ckeditor5-core/src/command';
```
Comment (rshipp) [inline app/javascript/ckeditor/insertcontentpartpreviewcommand.js]: js modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#insertproductpreviewcommandjs

File: app/javascript/components/ContentPartList.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Comment (rshipp) [inline app/javascript/components/ContentPartList.js]: js modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#productlistjs

File: app/javascript/components/ContentPartPreview.js [hunk 6]
```diff
@@ -0,0 +1,23 @@
+import React from 'react';
```
Comment (rshipp) [inline app/javascript/components/ContentPartPreview.js]: js modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#productpreviewjs

File: app/views/contents/_form.html.erb [hunk 7]
```diff
@@ -20,7 +20,25 @@
 
   <div class="field">
     <%= form.label :body %>
-    <%= react_component('ContentEditor', { data: content.body }) %>
+    <%= react_component('ContentEditor', { data: content.body,
+        contentParts: [
```
Comment (rshipp) [inline app/views/contents/_form.html.erb]: we'd read this `contentParts` thing from the database, this hardcoded list is just for the demo

File: app/javascript/components/ContentEditor.js [hunk 8]
```diff
@@ -1,46 +1,221 @@
...
+            'tableRow',
+            'mergeTableCells'
+        ]
+    },
+    // This value must be kept in sync with the language defined in webpack.config.js.
+    language: 'en'
 };
 
 class ContentEditor extends Component {
+    constructor( props ) {
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: most of this is modified from here https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#appjs, and rebased on top of the balloon-block editor build.

File: app/javascript/components/ContentEditor.js [hunk 9]
```diff
@@ -1,46 +1,221 @@
 import React, { Component } from 'react';
+import ReactDOM from 'react-dom';
 import CKEditor from '@ckeditor/ckeditor5-react';
 
+// The official CKEditor 5 instance inspector. It helps understand the editor view and model.
+import CKEditorInspector from '@ckeditor/ckeditor5-inspector';
+
 // NOTE: Use the editor from source (not a build)!
-import ClassicEditor from '@ckeditor/ckeditor5-editor-classic/src/classiceditor';
+import BalloonEditor from '@ckeditor/ckeditor5-editor-balloon/src/ballooneditor';
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: we can actually use arbitrarily many of these! I'm just using Balloon as the base for now, and adding custom Block stuff on top of it, but we can change it later if we want.

File: app/javascript/components/ContentEditor.js [hunk 10]
```diff
@@ -1,46 +1,221 @@
...
+    Link,
+    List,
+    MediaEmbed,
+    Paragraph,
+    PasteFromOffice,
+    Table,
+    TableToolbar,
+    UploadAdapter,
+
+    ContentPartPreviewEditing
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: this is how we tie in custom commands

File: app/javascript/components/ContentEditor.js [hunk 11]
```diff
@@ -1,46 +1,221 @@
...
+            // the user types and modifies the content of the editor.
+            editorData: props.data,
+        };
+
+        // The configuration of the <CKEditor> instance.
+        this.editorConfig = {
+            // The configuration of the ContentParts plugin. It specifies a function that will allow
+            // the editor to render a React <ContentPartPreview> component inside a contentPart widget.
+            contentParts: {
+                contentPartRenderer: ( id, domElement ) => {
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: this is the most important part! this function lets us insert `ContentPart` elements into the editor.

File: app/javascript/components/ContentEditor.js [hunk 12]
```diff
@@ -1,46 +1,221 @@
...
+        super( props );
+
+        // A place to store the reference to the editor instance created by the <CKEditor> component.
+        // The editor instance is created asynchronously and is only available when the editor is ready.
+        this.editor = null;
+
+        this.state = {
+            // The initial editor data. It is bound to the editor instance and will change as
+            // the user types and modifies the content of the editor.
+            editorData: props.data,
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: load the `data` property passed in from the view

File: app/javascript/components/ContentEditor.js [hunk 13]
```diff
@@ -1,46 +1,221 @@
...
-                        console.log( 'Blur.', editor );
-                    } }
-                    onFocus={ ( event, editor ) => {
-                        console.log( 'Focus.', editor );
-                    } }
+            <div className="demo" key="content-editor">
+                <div className="demo__offer-editor" key="offer-editor">
+                    <h3>Content editor</h3>
+                    <CKEditor
+                        editor={BalloonEditor}
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: this is where we set Balloon as the base

File: app/javascript/components/ContentEditor.js [hunk 14]
```diff
@@ -1,46 +1,221 @@
...
+                        data={this.state.editorData}
+                        config={this.editorConfig}
+                        onChange={this.handleEditorDataChange}
+                        onInit={this.handleEditorInit}
+                    />
+
+                    <h3>HTML View</h3>
+                    <textarea value={this.state.editorData} readOnly={true} name="content[body]"></textarea>
```
Comment (rshipp) [inline app/javascript/components/ContentEditor.js]: the `name` attribute here lets this textarea submit as part of the form, so we can create/update contents as normal

File: app.json [hunk 15]
```diff
@@ -7,8 +7,9 @@
       ],
       "addons": ["heroku-postgresql:in-dyno"],
       "scripts": {
+        "setup": "bundle exec rake db:migrate db:seed",
```
Comment (rshipp) [inline app.json]: this (hopefully) fixes the autodeploy for heroku review apps

File: app.json [hunk 16]
```diff
@@ -7,8 +7,9 @@
...
         "test-setup": "bundle exec rake webpacker:install",
-        "test": "bundle exec rspec"
+        "test": "bundle exec rspec && yarn test"
```
Comment (rshipp) [inline app.json]: run rspec and then jest

File: spec/javascript/components/ContentPartPreview.test.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
```
Comment (rshipp) [inline spec/javascript/components/ContentPartPreview.test.js]: an example Jest test for one of the non-ckeditor-based components

File: app/assets/stylesheets/application.scss (hunk 1 above)
Comment (sadleb) [reply app/assets/stylesheets/application.scss]: I think it's for compatibility with bootstrap: https://github.com/twbs/bootstrap-rubygem#a-ruby-on-rails If we're not going to use bootstrap or that issue is outdated at this point, this is prob fine. It used to cause issues with accessing Bootstrap's mixins and variables though.

**Pull request merged**